src/Std/Data/DHashMap/Internal/Defs.lean

```diff
@@ -0,0 +1,414 @@
+    `Internal.List.Model` via a lemma `operation_eq_operationₘ`.
+4. Verify the model implementation
+  * In `Internal.WF`, prove `operationₘ_eq_List.operation` (for access operations) or
+    `wfImp_operationₘ` and `toListModel_operationₘ`
+  * Immediately deduce the corresponding results `operation_eq_List.operation` or
+    `wfImp_operation`/`toListModel_operation` by combining the results from steps 3 and 4.
+  * If you added a new constructor to `Raw.WF` earlier, fix up the proof of `Raw.WF.out`.
+5. Connect `Raw` and `Raw₀`
+  * Write `Raw.operation_eq` and `Raw.operation_val` in `Internal.Raw`.
+6. Prove `Raw₀` versions of user-facing lemmas
+  * State all of the user-facing lemmas that you want in `Internal.RawLemmas`. This will generally
+    involve connecting the operation to ALL existing operations or deciding that there is nothing to
+    be said about a certain pair.
+  * Prove the corresponding results about lists in `List.Associative`
+  * Use the tactic provided in `Internal.RawLemmas` to deduce the `Raw₀` results from the list
+    results. You will need to hook some of the results you proved in step 4 into the tactic. You
+    might also have to prove that your list operation is invariant under permutation and add that to
+    the tactic.
+7. State and prove the user-facing lemmas
+  * Restate all of your lemmas for `DHashMap.Raw` in `DHashMap.Lemmas` and prove them using the
+    provided tactic after hooking in your `operation_eq` and `operation_val` from step 5.
+  * Restate all of your lemmas for `DHashMap` in `DHashMap.Lemmas` and prove them by reducing to
+    `Raw₀`.
+  * Restate all of your lemmas for `HashMap.Raw` in `HashMap.Lemmas` and prove them by reducing to
+    `DHashMap.Raw`.
+  * Restate all of your lemmas for `HashMap` in `HashMap.Lemmas` and prove them by reducing to
+    `DHashMap`.
+  * Restate all of your lemmas for `HashSet.Raw` in `HashSet.Lemmas` and prove them by reducing to
+    `DHashSet.Raw`.
+  * Restate all of your lemmas for `HashSet` in `HashSet.Lemmas` and prove them by reducing to
+    `DHashSet`.
+
+This sounds like a lot of work (and it is if you have to add a lot of user-facing lemmas), but the
+framework is set up in such a way that each step is really easy and the proofs are all really short
+and clean.
+-/
+
+set_option autoImplicit false
+
+universe u v w
+
+variable {α : Type u} {β : α → Type v} {δ : Type w} {m : Type w → Type w} [Monad m]
+
+namespace Std
+
+namespace DHashMap.Internal
+
+@[inline] private def numBucketsForCapacity (capacity : Nat) : Nat :=
+  -- a "load factor" of 0.75 is the usual standard for hash maps
+  capacity * 4 / 3
+
+def toListModel (buckets : Array (AssocList α β)) : List ((a : α) × β a) :=
+  buckets.data.bind AssocList.toList
+
+@[inline] def computeSize (buckets : Array (AssocList α β)) : Nat :=
+  buckets.foldl (fun d b => d + b.length) 0
+
+abbrev Raw₀ (α : Type u) (β : α → Type v) :=
```

Comment:
```suggestion
/--
This is an internal implementation detail of the DHashMap library and should not be used externally.
-/
abbrev Raw₀ (α : Type u) (β : α → Type v) :=
```

Reply:
My philosophy with these internal types was: they must live in the `Internal` folder and they must not be reachable in one step via go-to-definition from user case. So if a user finds themselves looking at this definition, they should already know that they're looking at an implementation detail. Do you still think it would make sense to annotate all the types in the `Internal` folders?

Reply:
I'm not sure, to be honest.

The concern I have is that they might show up in a proof state or the infoview, and get followed from there.

On the balance, I think it's better to have it on each type, because users do all sorts of things we don't expect, and a bit of redundant communication is better than a misunderstanding. After all, the goal is to keep internals internal so that user code doesn't break; it's not to have a thing to point at and say "I told you so" if it breaks and they complain.

But this is an opinion I hold lightly.

Reply:
Okay, I've put a "nothing to see here" docstring on every definition for now.